### PR TITLE
Limit concurrent requests to endorser/deliver services

### PR DESCRIPTION
**Pull request by wenjianqiao**
Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

#### Type of change
- Improvement (improvement to code, performance, etc)

#### Description

Add new config data to set limits for concurrent requests to endorser
and deliver services. Create semaphores using the configurable limits.
Implement grpc interceptors (for unary and stream) to acquire
a semaphore and return an error when the semaphore limit is exceeded.

#### Related issues
https://jira.hyperledger.org/browse/FAB-14761

#### Release Note
TBD

### CODE REVIEW

**Review by yacovm — DISMISSED** [review]

**Comment by yacovm** [comment]
An integration test where we have an artificial limit of 1 deliver stream can be nice.

File: internal/peer/node/throttle_grpc.go
Comment (sykesm) [inline internal/peer/node/throttle_grpc.go]: Please: No package level state. You have a function that returns it and . one consumer.

File: internal/peer/node/throttle_grpc.go
Comment (sykesm) [inline internal/peer/node/throttle_grpc.go]: Why do we need an interface? You're constructing a concrete type in the unexported constructor and have no tests that replace it with a mock.

File: internal/peer/node/throttle_grpc.go
Comment (sykesm) [inline internal/peer/node/throttle_grpc.go]: Seems fragile, but works.

**Review by sykesm — DISMISSED** [review]
I don't like doing this but duplicating a struct and a collection of tests instead of adding a method doesn't make a lot of sense to me. Combine that with the unnecessary interface and package level state and I think it warrants changes before merging.

File: internal/peer/node/throttle_grpc.go
Comment (yacovm) [reply internal/peer/node/throttle_grpc.go]: oh I totally missed this...

File: common/semaphore/throttle_semaphore.go
Comment (wenjianqiao) [reply common/semaphore/throttle_semaphore.go]: updated message

File: common/semaphore/throttle_semaphore.go
Comment (wenjianqiao) [reply common/semaphore/throttle_semaphore.go]: Thank you Matt for the comment. Have updated code.

File: internal/peer/node/throttle_grpc.go
Comment (wenjianqiao) [reply internal/peer/node/throttle_grpc.go]: Need the package level var because it is used by unaryThrottleInterceptor and streamThrottleInterceptor, but it cannot be passed in as a parameter.

File: internal/peer/node/throttle_grpc.go
Comment (wenjianqiao) [reply internal/peer/node/throttle_grpc.go]: Forgot to remove - it was there from an old (different) approach. Removed now.

File: internal/peer/node/throttle_grpc.go
Comment (wenjianqiao) [reply internal/peer/node/throttle_grpc.go]: Figured out a way to pass semaphores as a parameter and removed the package level var.

File: internal/peer/node/throttle_grpc.go
Comment (cendhu) [inline internal/peer/node/throttle_grpc.go]: Given that we are applying limits per service instead of per rpc call, can we use only the service name such as `/protos.Endorser` and `/protos.Deliver` as the key in the semaphore map? This will simply the above lines and make it less fragile whenever a new rpc is being added to a service. At least to preserve backward compatibility, I assume that we won't change the service name itself. 

If we make the above changes, in the interceptor, we can introduce a function called `getServiceName(info.MethodName)` to get the service name alone to do the lookup in the semaphore map.

Also, please add a comment stating that these services are defined in `fabric-protos` and `fabric-protos-go` repositories. It would help improve the readability.

File: core/peer/config.go
Comment (wenjianqiao) [reply core/peer/config.go:115]: The removal of qscc limit (and related code) will be done in a separate PR.

File: common/semaphore/semaphore.go
Comment (sykesm) [reply common/semaphore/semaphore.go:8]: We don't need a weighted semaphore so the additional dependency isn't warranted. The addition of `TryAcquire` without a context was all that was necessary here (without a context).

**Review by sykesm — COMMENTED** [review]
Merging the two implementations addressed my biggest concern.

File: internal/peer/node/throttle_grpc.go
Comment (wenjianqiao) [reply internal/peer/node/throttle_grpc.go]: @cendhu In this case, AcquireNow may return an error due to context canceled. Therefore, we are not sure if the error is too many request unless we introduce a new error type.

File: internal/peer/node/throttle_grpc.go
Comment (wenjianqiao) [reply internal/peer/node/throttle_grpc.go]: I thought about the same, but wasn't sure if it was better to add parsing to method name. Given your comment and Matt's comment, will parse the method name to get rpc name.

File: core/peer/config.go
Comment (wenjianqiao) [reply core/peer/config.go]: To maintain consistency, leave it as is unless the config params are changed in core.yaml.

File: sampleconfig/core.yaml
Comment (wenjianqiao) [reply sampleconfig/core.yaml]: Will discuss offline.

File: internal/peer/node/throttle_grpc.go
Comment (wenjianqiao) [reply internal/peer/node/throttle_grpc.go]: Rethinking about it, decide to remove context from AcquireNow so that it only returns error when the buffer is full. Context cancellation will be caught at grpc layer anyway. Error message is updated accordingly.

File: internal/peer/node/throttle_grpc.go
Comment (wenjianqiao) [reply internal/peer/node/throttle_grpc.go]: Agree, code updated.

File: sampleconfig/core.yaml
Comment (wenjianqiao) [reply sampleconfig/core.yaml:431]: Done

File: internal/peer/node/throttle_grpc.go
Comment (wenjianqiao) [reply internal/peer/node/throttle_grpc.go]: Renamed

File: internal/peer/node/throttle_grpc.go
Comment (wenjianqiao) [reply internal/peer/node/throttle_grpc.go]: Renamed both file name and method names.

File: common/semaphore/semaphore.go
Comment (cendhu) [reply common/semaphore/semaphore.go:8]: Got it. Thanks. As the code for both old and new approaches present in these files, I got confused and was wondering why to add 150+ lines of code rather than importing an existing package. Now, I realized that most existing code in these files would be removed and we just need `TryAcquire` without the context. Hence, I agree that it doesn't warrant a new package dependency.

File: common/semaphore/semaphore.go
Comment (cendhu) [inline common/semaphore/semaphore.go]: Nit. 
```suggestion
func (s Semaphore) TryAcquire() bool {
```
As in sync package, we can name it `TryAcquire()` and return `bool` instead of an `error`. I don't have a strong opinion on this and hence, I leave it to you.

**Review by cendhu — COMMENTED** [review]
Thanks, @wenjianqiao, for the changes.

File: common/semaphore/semaphore.go
Comment (wenjianqiao) [reply common/semaphore/semaphore.go]: Changed to TryAcquire.

**Comment by cendhu** [comment]
Thanks, Wenjian, for the changes. Looks good to me.

**Review by manish-sethi — COMMENTED** [review]
LGTM - Just one minor comment. Also, I guess that you can remove the qscc limit variable also in this PR. However, it's fine with me if you prefer to do that in a separate PR.

File: internal/peer/node/grpc_limiters.go
Comment (wenjianqiao) [reply internal/peer/node/grpc_limiters.go]: Thanks Manish. Code updated. 
Once this PR is merged, will push a new PR to remove qscc limit and related code.

**Review by manish-sethi — APPROVED** [review]

**Pull request merged**
